Migrate

ruihe774 opened the pull request.

rchl committed:
Initial commit

rchl committed:
Create initial chokidar file watcher (#1)

rchl committed:
Clarify usage

rchl committed:
fix grammar

rchl committed:
remove unnecessary and invalid flake8 ignore rules

rchl committed:
adapt for the new version of lsp_utils (#3)

rchl committed:
adapt to the new version of lsp_utils fixup

rchl committed:
feat: batch notification events (#6)

predragnikolic committed:
"Add .python-version file"

ruihe774 committed:
remove chokidar

ruihe774 committed:
add rust implementation

ruihe774 committed:
use glob

ruihe774 committed:
update readme stuff

ruihe774 committed:
use notify_debouncer_full

ruihe774 committed:
add normalize_events

ruihe774 committed:
add CI

ruihe774 committed:
do not use Arc

ruihe774 committed:
implement unwatch

ruihe774 committed:
add parent_process_watchdog for windows

ruihe774 committed:
enter efficiency mode in windows

ruihe774 committed:
enter_efficiency_mode for linux

ruihe774 committed:
fix parent_process_watchdog on win32

ruihe774 committed:
fix unwatching on windows & macos

ruihe774 committed:
fix canonicalize on windows

ruihe774 committed:
normalize sep on windows

ruihe774 committed:
add a compile error guard

ruihe774 committed:
do not need file id cache

ruihe774 committed:
do not need to emit modify after create

ruihe774 committed:
fix lifetime

ruihe774 committed:
require_literal_separator & require_literal_leading_dot

ruihe774 committed:
ci

ruihe774 committed:
support prebuilt

ruihe774 committed:
re-add lsp_utils

ruihe774 committed:
fix permission

ruihe774 committed:
update README.md

ruihe774 committed:
disable log in release

ruihe774 committed:
use my branch of rustix

rchl commented on .gitattributes:
```diff
@@ -0,0 +1,4 @@
+.github/ export-ignore
+.gitignore export-ignore
+pyrightconfig.json export-ignore
+tox.ini export-ignore
```
Can also include: `Cargo.lock`, `Cargo.toml`, `src/`.

Excluding here ensures that published package won't include those unnecessary files.

ruihe774 committed:
make crate private

ruihe774 committed:
ignore rust code in gitattributes

ruihe774 committed:
update dep

rchl committed:
cleanup dead code

rchl commented:
I've noticed some issue. When saving a watched file, it includes the same file twice in the notification.

> :: [22:19:33.863]  -> LSP-pyright workspace/didChangeWatchedFiles: {'changes': [{'uri': 'file:///usr/local/workspace/sublime-packages/LSP-file-watcher-rust/watcher.py', 'type': <FileChangeType.Changed: 2>}, {'uri': 'file:///usr/local/workspace/sublime-packages/LSP-file-watcher-rust/watcher.py', 'type': <FileChangeType.Changed: 2>}]}

For comparison, here is the event triggered by the original chokidar watcher:

> :: [22:21:28.364]  -> LSP-pyright workspace/didChangeWatchedFiles: {'changes': [{'uri': 'file:///usr/local/workspace/sublime-packages/LSP-file-watcher-rust/watcher.py', 'type': <FileChangeType.Changed: 2>}]}

rchl committed:
rename to rust-watcher

rchl commented:
Hey, is that something that you are planning to address? I think that this is a blocking issues.

ruihe774 commented:
Sorry for my late reply. I did not notice your previous comment. (Maybe because of an unread notification. Sorry.)

Personally I don't think it's a problem. My implementation does not mean to be a 1:1 replicate. Triggering change events twice is not a serious issue giving 1) the LSP protocol does not define the behavior strictly, and 2) the LSP server will still reload the changed file as expected.

My guess is that the editor saves the file by replacing the original file with an updated file. On some platforms, the 1) delete 2) rename events are not consolidated into one single event. I cannot reproduce the doubled change events in Linux.

The upstream `notify` crate does not guarantee a strict cross-platform consistency. I have to admit that this kind of inconsistencies of events can happen. However, I don't think they can lead to an unexpected inconsistency between files cached in the LSP server and in the workspace. As `notify` is used in widely used projects such as `cargo` and `deno`, I think it's mature enough.

If you think a 1:1 behavior is required, I can try to achieve that at the price of code complexity. I think the current behavior is acceptable, and we can wait for bug reports.

rchl commented:
I don't think it has to have the same behavior. I just saw it and thought that it looks like a bug that could negatively affect performance of servers since those might do the same work multiple times if those don't do any de-duplication themselves. I don't know how serious this is but if during branch changes it happens for every file that gets changed then it could be significant. I will try to investigate it more when I have some time.

ruihe774 commented:
> I will try to investigate it more when I have some time.

I will also look into this in my spare time.

rchl commented on src/main.rs:
```diff
@@ -0,0 +1,380 @@
+    for event in events {
+        let event_type = match event.kind {
+            notify::EventKind::Create(_) => EventType::Create,
+            notify::EventKind::Modify(_) => EventType::Change,
```
The reason for double notification on editing a file is the fact that we are being notified twice with two different event attributes - once with `Modify(Metadata(Any))` and once with `Modify(Data(Content))`:

```
event: Event { kind: Modify(Metadata(Any)), paths: ["/usr/local/workspace/sublime-packages/LSP-file-watcher-rust/src/main.rs"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None }
event: Event { kind: Modify(Data(Content)), paths: ["/usr/local/workspace/sublime-packages/LSP-file-watcher-rust/src/main.rs"], attr:tracker: None, attr:flag: None, attr:info: None, attr:source: None }
```

I believe that the correct thing to do would be to filter out any non-Content ones since LSP only wants to get content notification changes.

```suggestion
            notify::EventKind::Modify(ModifyKind::Data(DataChange::Content)) => EventType::Change,
```

(also requires adding some `use` statements)

ruihe774 replied:
I tested chokidar and found it reported metadata changes made with e.g. `touch` as well. So the problem was not not ignoring metadata changes but was not merging events. Fixed in 5eb580c. Please have a look.

rchl commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,88 @@
+          name: ${{ env.OS_NAME }}-arm64
+          path: target/aarch64-${{ env.TRIPLE_SUFFIX }}/release/rust-watcher
+        if: ${{ env.OS_NAME == 'linux' }}
+  push:
+    name: Push artifacts
+    needs: build
+    if: ${{ (github.event_name == 'push' || github.event_name == 'workflow_dispatch') && needs.build.result == 'success' }}
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v4
+      - name: Setup repo
+        run: |
+          git config user.name `git log --format=%an -1`
+          git config user.email `git log --format=%ae -1`
+          git checkout --orphan built
+      - name: Download artifacts
+        uses: actions/download-artifact@v4
+      - name: Commit repo
+        run: |
+          find -name 'rust-watcher*' -exec chmod +x {} +
+          git add --all
+          git commit -m "Built artifacts for ${{ github.sha }}"
+      - name: Fetch newest commit
+        run: |
+          git fetch origin
+          echo NEWEST_COMMIT=`git rev-parse origin/main` >> $GITHUB_ENV
+      - name: Push repo
+        run: git push -f origin built
+        if: ${{ env.NEWEST_COMMIT == github.sha }}
```
I don't quite see the code here that would put the binary in the correct place (`/osx-universal2` for mac, for example). Am I just missing?

EDIT: I guess on uploading artifact the `name` is set so maybe it ends up correct in the end?

ruihe774 replied:
Yeah, I think so. Anyway it works.

rchl commented on src/main.rs:
```diff
@@ -0,0 +1,380 @@
+                .collect()
+        };
+
+        let prefixes: Vec<_> = make_absolute_paths(&cwd, &req.patterns).collect();
```
This is not correct.

Example:
cwd: `/usr/local/foo/`
patterns: `['**/biome.json']`

This will create a prefix like `usr/local/foo/**/biome.json` which will then later fail `starts_with` check against the path which means it will fail to notify about this change.

rchl replied:
Actually the check that uses it uses negation. What's really the purpose of that check anyway?

But still there is some issue in filtering that I'm seeing so I need to see what exactly happens...

EDIT: Yeah, it's [that check](https://github.com/sublimelsp/LSP-file-watcher-rust/blob/82870b3c3f8404fe086939c622e2e98a53d16623/src/main.rs#L281-L284) that goes wrong in this case.

ruihe774 replied:
The logic is: if the path is matched by either globes or prefixes, we report it. Patterns like `/usr/local/foo/**/biome.json` is stored both in `patterns` and `prefixes`; we don't check whether it's a glob or prefix. We just report the path if either matches.

rchl replied:
The case I was having issue with still doesn't work.

Matching
`/usr/local/workspace/github/typescript-language-server/biome.json`
against
`/usr/local/workspace/github/typescript-language-server/**/biome.{json,jsonc}`
doesn't match.

(Perhaps I have simplified it too much when reporting it initially. Maybe it's the `{json,jsonc}` part that is not supported?)

ruihe774 replied:
Hmm, it's because we only monitor files underneath `cwd`. Yeah, chokidar supports monitoring patterns outside `cwd` but we currently do not. The Python script uses the name `root_path` for it so I thought it was the root we should watch. Are there any cases that LSP clients requesting watching outside it?

rchl replied:
No, the path is within the monitored directory. It's the pattern itself that is not supported.

A very old issue asking for it to be implemented https://github.com/rust-lang/glob/issues/2

ruihe774 replied:
Fixed in 9951ad0

ruihe774 committed:
move out parent_died

ruihe774 committed:
update deps

ruihe774 committed:
use one watcher per registration

ruihe774 committed:
fix event normalization

ruihe774 committed:
merge reports

ruihe774 committed:
Revert "fix event normalization"

This reverts commit 6b2eb608ec90afbb4e6d1c1180e53225df05862b.

ruihe774 committed:
do not use removed CI images

ruihe774 committed:
fix warning

rchl commented on watcher.py:
```diff
@@ -0,0 +1,188 @@
+        log('ERROR: {}'.format(message))
+
+    def on_transport_close(self, exit_code: int, exception: Optional[Exception]) -> None:
+        self._end_process(exception)
```
Note that I found an issue in this code in the chokidar watcher and I'm fixing it in https://github.com/sublimelsp/LSP-file-watcher-chokidar/pull/30

I will apply the same fix here since it's a copy more or less and has the same bug.

rchl committed:
fix issue with new process being ended accidentally

rchl committed:
simplify

rchl committed:
fix

rchl committed:
don't log on successful exit

ruihe774 committed:
use globset

rchl commented:
I've tested it by:
1. creating biome.json file
2. (re)saving it
3. deleting it.


<details>

<summary>Event log</summary>

```
<create biome.json>

 event: Event {
     kind: Create(
         File,
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Create(
         File,
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Metadata(
             Extended,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Metadata(
             Extended,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }

Notify: [{'uri': 'file:///usr/local/workspace/github/typescript-language-server/biome.json', 'type': <FileChangeType.Created: 1>}]

Starting watcher for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Kinds: ['create', 'change', 'delete']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']

<save biome.json>

 event: Event {
     kind: Modify(
         Metadata(
             Any,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Metadata(
             Any,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Data(
             Content,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Data(
             Content,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }

Notify: [{'uri': 'file:///usr/local/workspace/github/typescript-language-server/biome.json', 'type': <FileChangeType.Changed: 2>}]

Starting watcher for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Kinds: ['create', 'change', 'delete']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']

<delete biome.json>

 event: Event {
     kind: Modify(
         Name(
             Any,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Metadata(
             Any,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Data(
             Content,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Name(
             Any,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Metadata(
             Any,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }
 event: Event {
     kind: Modify(
         Data(
             Content,
         ),
     ),
     paths: [
         "/usr/local/workspace/github/typescript-language-server/biome.json",
     ],
     attr:tracker: None,
     attr:flag: None,
     attr:info: None,
     attr:source: None,
 }

Notify: [{'uri': 'file:///usr/local/workspace/github/typescript-language-server/biome.json', 'type': <FileChangeType.Changed: 2>}]

Starting watcher for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Kinds: ['create', 'change', 'delete']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
```

</details>

Notice it reports "change" after the file was deleted instead of "delete".
(The logged events are captured raw before `normalize_events` here: https://github.com/ruihe774/LSP-file-watcher-rust/blob/9951ad062e24cae2ac4c0c41ae52a01d9c7aebc7/src/main.rs#L263-L265)

(also note that the server re-registers the watcher after every change but that probably is not relevant)

rchl commented:
~~Also the events seems to double every time the server re-registers the watcher so I think there is also a problem with the old watchers not being unregistered.~~

EDIT: Ignore this part. Was slightly confused. The issue is that the events are always double the expected amount since Biome incorrectly registers two watchers.

ruihe774 commented:
> I've tested it by:
> 
> 1. creating biome.json file
> 2. (re)saving it
> 3. deleting it.
> 
> Event log
> ```
> <create biome.json>
> 
>  event: Event {
>      kind: Create(
>          File,
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Create(
>          File,
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Metadata(
>              Extended,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Metadata(
>              Extended,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
> 
> Notify: [{'uri': 'file:///usr/local/workspace/github/typescript-language-server/biome.json', 'type': <FileChangeType.Created: 1>}]
> 
> Starting watcher for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Kinds: ['create', 'change', 'delete']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
> 
> <save biome.json>
> 
>  event: Event {
>      kind: Modify(
>          Metadata(
>              Any,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Metadata(
>              Any,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Data(
>              Content,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Data(
>              Content,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
> 
> Notify: [{'uri': 'file:///usr/local/workspace/github/typescript-language-server/biome.json', 'type': <FileChangeType.Changed: 2>}]
> 
> Starting watcher for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Kinds: ['create', 'change', 'delete']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
> 
> <delete biome.json>
> 
>  event: Event {
>      kind: Modify(
>          Name(
>              Any,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Metadata(
>              Any,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Data(
>              Content,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Name(
>              Any,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Metadata(
>              Any,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
>  event: Event {
>      kind: Modify(
>          Data(
>              Content,
>          ),
>      ),
>      paths: [
>          "/usr/local/workspace/github/typescript-language-server/biome.json",
>      ],
>      attr:tracker: None,
>      attr:flag: None,
>      attr:info: None,
>      attr:source: None,
>  }
> 
> Notify: [{'uri': 'file:///usr/local/workspace/github/typescript-language-server/biome.json', 'type': <FileChangeType.Changed: 2>}]
> 
> Starting watcher for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Kinds: ['create', 'change', 'delete']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
> ```
> 
> Notice it reports "change" after the file was deleted instead of "delete". (The logged events are captured raw before `normalize_events` here: https://github.com/ruihe774/LSP-file-watcher-rust/blob/9951ad062e24cae2ac4c0c41ae52a01d9c7aebc7/src/main.rs#L263-L265)
> 
> (also note that the server re-registers the watcher after every change but that probably is not relevant)

Yeah, I can reproduce it in my MacBook. I cannot see a remove event from `notify`. It seems to be macOS-specific issue with `notify`.

ruihe774 committed:
fix macOS event translation

ruihe774 committed:
Revert "fix macOS event translation"

This reverts commit fade2da8866216da0a9e867bcf93170ebaa70a5e.

ruihe774 committed:
try kqueue backend in macOS

ruihe774 commented:
Maybe it's because of https://github.com/notify-rs/notify/issues/729. I'm trying the kqueue backend and it works fine.

rchl commented:
Sorry for the delay. Got to testing it now and it doesn't really work for me:

```
LSP-file-watcher-rust: Starting watcher "id: 2" for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
LSP-file-watcher-rust: Starting watcher "id: 3" for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/.eslintr{c.js,c.cjs,c.yaml,c.yml,c,c.json}', '**/eslint.config.{js,mjs,cjs,ts,mts,cts}', '**/.eslintignore', '**/package.json']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
LSP-file-watcher-rust: ERROR: Request Register(Register { cwd: "/usr/local/workspace/github/typescript-language-server", events: [Create, Change, Delete], ignores: ["**/.svn/**", "**/.git/**", "**/.hg/**", "**/CVS/**", "**/.mypy_cache/**", "**/*.pyc", "**/*.pyo", "**/*.exe", "**/*.dll", "**/*.obj", "**/*.o", "**/*.a", "**/*.lib", "**/*.so", "**/*.dylib", "**/*.ncb", "**/*.sdf", "**/*.suo", "**/*.pdb", "**/*.idb", "**/.DS_Store", "**/*.class", "**/*.psd", "**/node_modules/**"], patterns: ["**/biome.{json,jsonc}", ".editorconfig", "**/.gitignore", "**/.ignore"], uid: 2 })
LSP-file-watcher-rust: Starting watcher "id: 4" for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/*']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
LSP-file-watcher-rust: ERROR: failed to watch on path: Error { kind: Generic("Too many open files (os error 24) about [\"/usr/local/workspace/github/typescript-language-server/node_modules/@typescript-eslint/typescript-estree/node_modules/minimatch/dist/commonjs/escape.d.ts\"]"), paths: [] }
LSP-file-watcher-rust: ERROR: Request Register(Register { cwd: "/usr/local/workspace/github/typescript-language-server", events: [Create, Change, Delete], ignores: ["**/.svn/**", "**/.git/**", "**/.hg/**", "**/CVS/**", "**/.mypy_cache/**", "**/*.pyc", "**/*.pyo", "**/*.exe", "**/*.dll", "**/*.obj", "**/*.o", "**/*.a", "**/*.lib", "**/*.so", "**/*.dylib", "**/*.ncb", "**/*.sdf", "**/*.suo", "**/*.pdb", "**/*.idb", "**/.DS_Store", "**/*.class", "**/*.psd", "**/node_modules/**"], patterns: ["**/.eslintr{c.js,c.cjs,c.yaml,c.yml,c,c.json}", "**/eslint.config.{js,mjs,cjs,ts,mts,cts}", "**/.eslintignore", "**/package.json"], uid: 3 })
LSP-file-watcher-rust: ERROR: failed to watch on path: Error { kind: Generic("Too many open files (os error 24) about [\"/usr/local/workspace/github/typescript-language-server/node_modules/@typescript-eslint/typescript-estree/node_modules/minimatch/dist/commonjs/escape.d.ts\"]"), paths: [] }
LSP-file-watcher-rust: ERROR: Request Register(Register { cwd: "/usr/local/workspace/github/typescript-language-server", events: [Create, Change, Delete], ignores: ["**/.svn/**", "**/.git/**", "**/.hg/**", "**/CVS/**", "**/.mypy_cache/**", "**/*.pyc", "**/*.pyo", "**/*.exe", "**/*.dll", "**/*.obj", "**/*.o", "**/*.a", "**/*.lib", "**/*.so", "**/*.dylib", "**/*.ncb", "**/*.sdf", "**/*.suo", "**/*.pdb", "**/*.idb", "**/.DS_Store", "**/*.class", "**/*.psd", "**/node_modules/**"], patterns: ["**/*"], uid: 4 })
LSP-file-watcher-rust: ERROR: failed to watch on path: Error { kind: Generic("Too many open files (os error 24) about [\"/usr/local/workspace/github/typescript-language-server/node_modules/@typescript-eslint/typescript-estree/node_modules/minimatch/dist/commonjs/escape.d.ts\"]"), paths: [] }
```

So there is a `Too many open files (os error 24)` error. I would understand that it might hit some system limit (that I can probably raise) but when using chokidar watcher everything works. Chokidar uses Node's native fs.watch API which also seems to use `kqueue` on Mac - it's mentioned in https://nodejs.org/docs/latest/api/fs.html#fswatchfilename-options-listener

rchl commented:
Note that there are multiple registrations for the same directory (from 3 different servers that are running simultaneously). Perhaps chokidar or Node.js de-duplicates those and only sets up one watcher but I always thought there is no such logic there.

ruihe774 commented:
> Sorry for the delay. Got to testing it now and it doesn't really work for me:
> 
> ```
> LSP-file-watcher-rust: Starting watcher "id: 2" for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/biome.{json,jsonc}', '.editorconfig', '**/.gitignore', '**/.ignore']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
> LSP-file-watcher-rust: Starting watcher "id: 3" for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/.eslintr{c.js,c.cjs,c.yaml,c.yml,c,c.json}', '**/eslint.config.{js,mjs,cjs,ts,mts,cts}', '**/.eslintignore', '**/package.json']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
> LSP-file-watcher-rust: ERROR: Request Register(Register { cwd: "/usr/local/workspace/github/typescript-language-server", events: [Create, Change, Delete], ignores: ["**/.svn/**", "**/.git/**", "**/.hg/**", "**/CVS/**", "**/.mypy_cache/**", "**/*.pyc", "**/*.pyo", "**/*.exe", "**/*.dll", "**/*.obj", "**/*.o", "**/*.a", "**/*.lib", "**/*.so", "**/*.dylib", "**/*.ncb", "**/*.sdf", "**/*.suo", "**/*.pdb", "**/*.idb", "**/.DS_Store", "**/*.class", "**/*.psd", "**/node_modules/**"], patterns: ["**/biome.{json,jsonc}", ".editorconfig", "**/.gitignore", "**/.ignore"], uid: 2 })
> LSP-file-watcher-rust: Starting watcher "id: 4" for directory "/usr/local/workspace/github/typescript-language-server". Pattern: ['**/*']. Ignores: ['**/.svn/**', '**/.git/**', '**/.hg/**', '**/CVS/**', '**/.mypy_cache/**', '**/*.pyc', '**/*.pyo', '**/*.exe', '**/*.dll', '**/*.obj', '**/*.o', '**/*.a', '**/*.lib', '**/*.so', '**/*.dylib', '**/*.ncb', '**/*.sdf', '**/*.suo', '**/*.pdb', '**/*.idb', '**/.DS_Store', '**/*.class', '**/*.psd', '**/node_modules/**']
> LSP-file-watcher-rust: ERROR: failed to watch on path: Error { kind: Generic("Too many open files (os error 24) about [\"/usr/local/workspace/github/typescript-language-server/node_modules/@typescript-eslint/typescript-estree/node_modules/minimatch/dist/commonjs/escape.d.ts\"]"), paths: [] }
> LSP-file-watcher-rust: ERROR: Request Register(Register { cwd: "/usr/local/workspace/github/typescript-language-server", events: [Create, Change, Delete], ignores: ["**/.svn/**", "**/.git/**", "**/.hg/**", "**/CVS/**", "**/.mypy_cache/**", "**/*.pyc", "**/*.pyo", "**/*.exe", "**/*.dll", "**/*.obj", "**/*.o", "**/*.a", "**/*.lib", "**/*.so", "**/*.dylib", "**/*.ncb", "**/*.sdf", "**/*.suo", "**/*.pdb", "**/*.idb", "**/.DS_Store", "**/*.class", "**/*.psd", "**/node_modules/**"], patterns: ["**/.eslintr{c.js,c.cjs,c.yaml,c.yml,c,c.json}", "**/eslint.config.{js,mjs,cjs,ts,mts,cts}", "**/.eslintignore", "**/package.json"], uid: 3 })
> LSP-file-watcher-rust: ERROR: failed to watch on path: Error { kind: Generic("Too many open files (os error 24) about [\"/usr/local/workspace/github/typescript-language-server/node_modules/@typescript-eslint/typescript-estree/node_modules/minimatch/dist/commonjs/escape.d.ts\"]"), paths: [] }
> LSP-file-watcher-rust: ERROR: Request Register(Register { cwd: "/usr/local/workspace/github/typescript-language-server", events: [Create, Change, Delete], ignores: ["**/.svn/**", "**/.git/**", "**/.hg/**", "**/CVS/**", "**/.mypy_cache/**", "**/*.pyc", "**/*.pyo", "**/*.exe", "**/*.dll", "**/*.obj", "**/*.o", "**/*.a", "**/*.lib", "**/*.so", "**/*.dylib", "**/*.ncb", "**/*.sdf", "**/*.suo", "**/*.pdb", "**/*.idb", "**/.DS_Store", "**/*.class", "**/*.psd", "**/node_modules/**"], patterns: ["**/*"], uid: 4 })
> LSP-file-watcher-rust: ERROR: failed to watch on path: Error { kind: Generic("Too many open files (os error 24) about [\"/usr/local/workspace/github/typescript-language-server/node_modules/@typescript-eslint/typescript-estree/node_modules/minimatch/dist/commonjs/escape.d.ts\"]"), paths: [] }
> ```
> 
> So there is a `Too many open files (os error 24)` error. I would understand that it might hit some system limit (that I can probably raise) but when using chokidar watcher everything works. Chokidar uses Node's native fs.watch API which also seems to use `kqueue` on Mac - it's mentioned in https://nodejs.org/docs/latest/api/fs.html#fswatchfilename-options-listener

Yes, it's a known issue with kqueue backend. FSEvents is more useable. We can watch the progress of https://github.com/notify-rs/notify/issues/729.

rchl commented:
I saw that there were some upstream fixes so I wonder if things can be improved now.

It would be nice to get this to the finish line but I'd like to first compare the final implementation's memory usage and file handle usage with chokidar to ensure that it's not performing worse.

rchl commented:
I found some more info on how chokidar (v3 - the one we are using) works on Mac - https://github.com/paulmillr/chokidar/issues/1452

It compares it to v5 which uses native node APIs and tends to also fail due to large number of descriptors. v3 appears to be using `fsevents` and is more reliable.

ruihe774 committed:
rewrite implemention

ruihe774 committed:
add clippy

ruihe774 committed:
shared watcher

ruihe774 committed:
fix ext matching

ruihe774 commented:
I just done a major refactor and trying to fix all observed issues. Please have a try.

ruihe774 committed:
fix windows building

ruihe774 committed:
add parity check

ruihe774 committed:
migrate to newer macos ci image

ruihe774 commented:
I wrote a parity checking script and it passed on all OSes in the CI

rchl commented:
In the meantime LSP has refactored transports so this code wouldn't work anymore.

I've pushed vendored transports to make it work again, same as I did for chokidar watcher.

rchl commented:
I don't have 100% reproducible way for it to happen but the rust-process is not shutting down for me often when closing ST (on mac). In that case it will also sit using 20% of CPU typically or sometimes even jump to 100%.

It seems like the process eventually closes but it can take a minute or two.

And I'm not even opening that big directories (not a lot of directories/files).
